src/ioctls/kvm_vec.rs

```diff
@@ -0,0 +1,394 @@
+    ///
+    pub fn as_ptr(&self) -> *const T {
+        self.as_original_struct()
+    }
+
+    /// Get a mutable pointer to the KVM struct so it can be passed to the kernel.
+    ///
+    pub fn as_mut_ptr(&mut self) -> *mut T {
+        self.as_mut_original_struct()
+    }
+
+    /// Returns a mut `Vec<F>` that contains all the elements.
+    /// It is important to call `mem::forget` after using this vector. Otherwise rust will destroy it.
+    ///
+    unsafe fn as_vec(&mut self) -> Vec<F> {
+        let entries_ptr = self.as_mut_original_struct().entries_mut().as_mut_ptr();
+        Vec::from_raw_parts(
+            entries_ptr,
+            self.as_mut_original_struct().len() as usize,
+            self.capacity as usize,
+        )
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_entries_slice(&self) -> &[F] {
+        let len = self.as_original_struct().len();
+        unsafe { self.as_original_struct().entries().as_slice(len as usize) }
+    }
+
+    /// Get the mutable elements slice so they can be modified before passing to the VCPU.
+    ///
+    pub fn as_mut_entries_slice(&mut self) -> &mut [F] {
+        let len = self.as_original_struct().len();
+        unsafe {
+            self.as_mut_original_struct()
+                .entries_mut()
+                .as_mut_slice(len as usize)
+        }
+    }
+
+    /// Reserves capacity for at least `additional` more `F` elements.
+    ///
+    fn reserve(&mut self, additional: usize) {
+        let current_kvm_array_len = self.kvm_array.len();
+        let required_kvm_array_len =
+            KvmVec::<T, F>::num_elements_to_kvm_array_len(self.capacity + additional);
+        let additional_kvm_array_len = required_kvm_array_len - current_kvm_array_len;
+
+        self.capacity += additional;
+
+        self.kvm_array.reserve(additional_kvm_array_len);
+        for _ in current_kvm_array_len..required_kvm_array_len {
+            self.kvm_array.push(T::default())
+        }
+    }
+
+    /// Appends an element to the end of the collection.
+    ///
+    pub fn push(&mut self, entry: F) -> Result<(), Error> {
+        if self.as_mut_original_struct().len() + 1 > T::max_len() {
+            return Err(Error::SizeLimitExceeded);
+        }
+
+        self.reserve(1);
+
+        let mut entries = unsafe { self.as_vec() };
+        entries.push(entry);
+        *self.as_mut_original_struct().len_mut() = entries.len() as u32;
+
+        mem::forget(entries);
+
+        Ok(())
+    }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    pub fn retain<P>(&mut self, f: P)
+    where
+        P: FnMut(&F) -> bool,
+    {
+        let mut entries = unsafe { self.as_vec() };
+        entries.retain(f);
+        *self.as_mut_original_struct().len_mut() = entries.len() as u32;
+        self.capacity = entries.len();
+
+        self.kvm_array.shrink_to_fit();
```

Comment:
I don't think this has any effect since the number of `T` elements in `kvm_array` has not changed (even if you've changed the underlying memory re-interpreted as a `Vec<F>` when the `entries.retain(f)` was done).

Reply:
True. The len of the kvm_array should be changed in order for this to have any effect. Fixed.